### PR TITLE
Multi challenge fixes

### PR DESCRIPTION
**Pull request by cornelinux**
I have some fixes here, that came up in a real life test with RADIUS.

### CODE REVIEW

File: tests/test_api_validate.py [hunk 1]
```diff
@@ -2876,6 +2876,152 @@ def test_00_registrationtoken_with_validity_period(self):
...
+            result = res.json['result']
+            self.assertTrue(result.get("value"))
+
+        # 1st authentication creates a PIN change challenge via challenge response
+        with self.app.test_request_context('/validate/check', method='POST',
+                                           data={"user": "cornelius",
+                                                 "pass": "test"}):
+            res = self.app.full_dispatch_request()
+            self.assertEqual(res.status_code, 200)
+            details = res.json['detail']
```
Comment (plettich) [inline tests/test_api_validate.py:2909]: Please also check that the result value is `False` so we don't miss any possible mis-configurations

File: tests/test_api_validate.py [hunk 2]
```diff
@@ -2876,6 +2876,152 @@ def test_00_registrationtoken_with_validity_period(self):
...
+            transaction_id = details.get("transaction_id")
+
+        with self.app.test_request_context('/validate/check', method='POST',
+                                           data={"user": "cornelius",
+                                                 "pass": self.valid_otp_values[1],
+                                                 "transaction_id": transaction_id}):
+            res = self.app.full_dispatch_request()
+            self.assertEqual(res.status_code, 200)
+            result = res.json['result']
+            details = res.json['detail']
```
Comment (plettich) [inline tests/test_api_validate.py:2919]: same here

File: tests/test_api_validate.py [hunk 3]
```diff
@@ -2876,6 +2876,152 @@ def test_00_registrationtoken_with_validity_period(self):
...
+            self.assertEqual("Please enter a new PIN", details.get("message"))
+
+        with self.app.test_request_context('/validate/check', method='POST',
+                                           data={"user": "cornelius",
+                                                 "pass": "newpin",
+                                                 "transaction_id": transaction_id}):
+            res = self.app.full_dispatch_request()
+            self.assertEqual(res.status_code, 200)
+            result = res.json['result']
+            details = res.json['detail']
```
Comment (plettich) [inline tests/test_api_validate.py:2930]: and here

File: tests/test_api_validate.py [hunk 4]
```diff
@@ -2876,6 +2876,152 @@ def test_00_registrationtoken_with_validity_period(self):
...
+            result = res.json['result']
+            self.assertTrue(result.get("value"))
+
+        # 1st authentication creates a PIN change challenge via challenge response
+        with self.app.test_request_context('/validate/check', method='POST',
+                                           data={"user": "cornelius",
+                                                 "pass": "test{0!s}".format(self.valid_otp_values[1])}):
+            res = self.app.full_dispatch_request()
+            self.assertEqual(res.status_code, 200)
+            result = res.json['result']
```
Comment (plettich) [inline tests/test_api_validate.py:2982]: Please check the result value here as well

File: tests/test_api_validate.py [hunk 5]
```diff
@@ -2876,6 +2876,152 @@ def test_00_registrationtoken_with_validity_period(self):
...
+            self.assertEqual("Please enter a new PIN", details.get("message"))
+
+        with self.app.test_request_context('/validate/check', method='POST',
+                                           data={"user": "cornelius",
+                                                 "pass": "newpin",
+                                                 "transaction_id": transaction_id}):
+            res = self.app.full_dispatch_request()
+            self.assertEqual(res.status_code, 200)
+            result = res.json['result']
+            details = res.json['detail']
```
Comment (plettich) [inline tests/test_api_validate.py:2994]: and here

**Review by plettich — CHANGES_REQUESTED** [review]

File: privacyidea/lib/challengeresponsedecorators.py [hunk 6]
```diff
@@ -131,7 +132,11 @@ def generic_challenge_response_reset_pin(wrapped_function, *args, **kwds):
                     # Verify if the PIN adheres to the PIN policies. This is always in the normal user context
                     g = options.get("g")
                     g.logged_in_user = {"role": SCOPE.USER}
-                    check_pin(g, args[1], token_obj.token.tokentype, args[2])
+                    if user_obj:
+                        if __name__ == '__main__':
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/challengeresponsedecorators.py:136]: This is interesting.
I can not remember adding this line of code! 
I will remove it

File: privacyidea/lib/challengeresponsedecorators.py [hunk 7]
```diff
@@ -131,7 +132,11 @@ def generic_challenge_response_reset_pin(wrapped_function, *args, **kwds):
...
+                            g.logged_in_user["username"] = user_obj.login
+                            g.logged_in_user["realm"] = user_obj.realm
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/challengeresponsedecorators.py:138]: In this case indeed ``check_pin`` is a policy that does not work on a user as an object to be acted on, but as a subject which is acting.

``check_pin`` is used if a user tries to change his PIN in the UI, i.e. if a user is logged in. The request does not contain a user_obj (like when an administrator enrolls a token **for** a user).

In the context of multi challenge we are in a /validate/check request, where the ``check_pin`` policy would not find a user, to check. So we need to assume in this context that the user (who is trying to authenticate via /validate/check) would actually be the "logged_in" **acting** user.

File: privacyidea/lib/challengeresponsedecorators.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/challengeresponsedecorators.py]: done in 956a506d

File: privacyidea/lib/challengeresponsedecorators.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/lib/challengeresponsedecorators.py]: Could You add a comment indicating that usually only logged in users were able to change the pin and we therefore need this construct? Would make thinks much clearer later.

**Review by plettich — COMMENTED** [review]
check my comments

File: privacyidea/lib/challengeresponsedecorators.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/challengeresponsedecorators.py]: done in 95335454

File: tests/test_api_validate.py (hunk 1 above)
Comment (cornelinux) [reply tests/test_api_validate.py:2911]: done in f7589747

File: tests/test_api_validate.py (hunk 2 above)
Comment (cornelinux) [reply tests/test_api_validate.py:2921]: done in f7589747

File: tests/test_api_validate.py (hunk 3 above)
Comment (cornelinux) [reply tests/test_api_validate.py:2934]: done in f7589747

File: tests/test_api_validate.py (hunk 4 above)
Comment (cornelinux) [reply tests/test_api_validate.py:2986]: done in f7589747

File: tests/test_api_validate.py (hunk 5 above)
Comment (cornelinux) [reply tests/test_api_validate.py:2999]: done in f7589747

Comment (plettich) [reply tests/test_api_validate.py:2999]: Ähm, nope...

**Review by plettich — CHANGES_REQUESTED** [review]

**Comment by cornelinux** [comment]
@plettich too many occurances 8feace72

File: tests/test_api_validate.py (hunk 5 above)
Comment (cornelinux) [reply tests/test_api_validate.py:3000]: 8feace72

Comment (plettich) [reply tests/test_api_validate.py:3000]: ;-)

**Review by plettich — APPROVED** [review]
Looks good, i'll merge when travis has finished.

**Pull request merged**